addons/pos_adyen/controllers/main.py

```diff
@@ -18,18 +21,61 @@ def notification(self):
-        payment_method_sudo = request.env['pos.payment.method'].sudo().search([('adyen_terminal_identifier', '=', terminal_identifier)], limit=1)
-        pos_session_id = int(data["SaleToPOIResponse"]["PaymentResponse"]["SaleData"]["SaleTransactionID"]["TransactionID"].split("-")[1])
-        pos_session_sudo = request.env["pos.session"].sudo().browse(pos_session_id)
 
-        if payment_method_sudo:
-            # These are only used to see if the terminal is reachable,
-            # store the most recent ID we received.
-            if data['SaleToPOIResponse'].get('DiagnosisResponse'):
-                payment_method_sudo.adyen_latest_diagnosis = data['SaleToPOIResponse']['MessageHeader']['ServiceID']
-            else:
-                payment_method_sudo.adyen_latest_response = json.dumps(data)
-                request.env["bus.bus"].sudo()._sendone(pos_session_sudo._get_bus_channel_name(), "ADYEN_LATEST_RESPONSE", pos_session_sudo.config_id.id)
-        else:
-            _logger.error('received a message for a terminal not registered in Odoo: %s', terminal_identifier)
+        # Temporary fix, this whole system should be changed.
+        msg_header = data['SaleToPOIResponse'].get('MessageHeader')
+        if not msg_header \
+            or msg_header.get('ProtocolVersion') != '3.0' \
+            or msg_header.get('MessageClass') != 'Service' \
+            or msg_header.get('MessageType') != 'Response' \
+            or msg_header.get('MessageCategory') != 'Payment' \
+            or not msg_header.get('POIID'):
+            _logger.warning('Received an unexpected Adyen notification')
+            return
+
+        terminal_identifier = msg_header['POIID']
+        adyen_pm_sudo = request.env['pos.payment.method'].sudo().search([('adyen_terminal_identifier', '=', terminal_identifier)], limit=1)
+        if not adyen_pm_sudo:
+            _logger.warning('Received an Adyen event notification for a terminal not registered in Odoo: %s', terminal_identifier)
+            return
+
+        try:
+            adyen_additional_response = data['SaleToPOIResponse']['PaymentResponse']['Response']['AdditionalResponse']
+            pos_hmac = PosAdyenController._get_additional_data_from_unparsed(adyen_additional_response, 'metadata.pos_hmac')
+
+            msg_header = data['SaleToPOIResponse']['MessageHeader']
+            if not pos_hmac or not consteq(pos_hmac, adyen_pm_sudo._get_hmac(msg_header['SaleID'], msg_header['ServiceID'], msg_header['POIID'], data['SaleToPOIResponse']['PaymentResponse']['SaleData']['SaleTransactionID']['TransactionID'])):
+                _logger.warning('Received invalid Adyen event notification (invalid hmac): \n%s', pprint.pformat(data))
+                return
+
+            pos_hmac_metadata_raw = 'metadata.pos_hmac='+pos_hmac
+            safe_additional_response = adyen_additional_response.replace('&'+pos_hmac_metadata_raw, '').replace(pos_hmac_metadata_raw, '')
```

Comment:
Yes it is necessary because the Adyen response is sent to the cashier (POS frontend web interface). If the answer is an error like "payment aborted", then the cashier could reuse the same hmac with the same information and replace the error with a fake successful payment notification.